test/e2e/firewall.go

```diff
@@ -0,0 +1,411 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/service"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	gcecloud "k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	compute "google.golang.org/api/compute/v1"
+)
+
+const (
+	firewallTimeoutDefault  = 3 * time.Minute
+	firewallTestHttpTimeout = time.Duration(1 * time.Second)
+	// Set ports outside of 30000-32767, 80 and 8080 to avoid being whitelisted by the e2e cluster
+	firewallTestHttpPort = int32(29999)
+	firewallTestUdpPort  = int32(29998)
+)
+
+var _ = framework.KubeDescribe("Firewall rule", func() {
+	var firewall_test_name = "firewall-test"
+	f := framework.NewDefaultFramework(firewall_test_name)
+
+	// This test takes around 3 minutes to run
+	It("[slow] [serial] should create valid firewall rules for LoadBalancer type service", func() {
+		framework.SkipUnlessProviderIs("gce")
+		c := f.ClientSet
+		ns := f.Namespace.Name
+		cloudConfig := framework.TestContext.CloudConfig
+		gceCloud := framework.TestContext.CloudConfig.Provider.(*gcecloud.GCECloud)
+		// This source ranges is just used to examine we have the same things on LB firewall rules
+		firewallTestSourceRanges := []string{"0.0.0.0/1", "128.0.0.0/1"}
+
+		// OnlyLocal service is needed to examine which exact nodes the requests are being forward to by the Load Balancer on GCE
+		By("Creating a LoadBalancer type service with onlyLocal annotation")
+		serviceName := "firewall-test-loadbalancer"
+		jig := NewServiceTestJig(c, serviceName)
+		svc := createOnlyLocalLoadBalancServiceWithPort(jig, ns, serviceName, firewallTestHttpPort, firewallTestSourceRanges)
+		defer func() {
+			Expect(c.Core().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
+		}()
+		svcExternalIP := svc.Status.LoadBalancer.Ingress[0].IP
+
+		By("Checking if service's firewall rules are correct")
+		Expect(examineSvcFirewallRules(gceCloud, svc)).NotTo(HaveOccurred())
+
+		By("Creating a daemon set of netexec pods ")
+		dsName := "firewall-test-daemon-set"
+		Expect(createNetexecDaemonSet(c, dsName, ns, jig.Labels, firewallTestHttpPort, firewallTestUdpPort, true)).NotTo(HaveOccurred())
+		defer func() {
+			deleteDaemonSet(f, dsName, ns, jig.Labels)
+		}()
+
+		By("Checking that daemon pods launch on every node of the cluster.")
+		Expect(wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnAllNodes(f, jig.Labels))).NotTo(HaveOccurred(),
+			"error waiting for daemon pod to start")
+		Expect(checkDaemonStatus(f, dsName)).NotTo(HaveOccurred())
+
+		// Send request from outside of the cluster because internal pods trying to reach external
```

Comment:
I realize this comment is not clear. Accessing from outside because e2e cluster has this rule:
```
e2e-default-internal-node 10.0.0.0/8 tcp:1-65535, udp:1-65535, icmp e2e-test-zihongz-minion
```
And it will allow internal traffic to reach any nodes.